KEP-2891: Simplified scheduler plugin config (#2899)
kubernetes/enhancements: keps/sig-scheduling/2891-simplified-config/README.md

However these "behind-the-scenes" endpoints must still be configured in the `KubeSchedulerConfiguration`
provided by the user for every plugin that uses them. For the average user this is an unnecessary
and confusing step -- they only care that the plugin works, not how it works, and so
internal processing steps are usually irrelevant to them.

Therefore, a way for users to simply toggle all of a plugin's extension points
(or a subset of the extension points as defined necessary by the plugin developer)
would remove complexity and the potential for misconfiguration.

### Goals

* Provide an option for simplified plugin config from the end-user's perspective
* Enable this config in all in-tree default scheduler plugins

### Non-Goals

* Refactor or introduce any breaking changes to the scheduling framework

## Proposal

A new extension point, `MultiPoint` (final name TBD) will be made available to
users and developers which handles plugin registration for multiple extension
points simultaneously.

From the user's perspective, this will be a new (optional) field in the scheduler's
config to enable or disable plugins across multiple extension points.

> ### User Stories (Optional)

Reviewer: One important point here is that this multipoint approach doesn't solve the problem of refactoring plugins to add new extension points since we will continue to support the fine grained configuration. We can only do this with a new CC version. Perhaps worth pointing that out somewhere?

Reviewer: I think we should consider fusing PreScore + Score, PreFilter + Filter and PreBind+Bind from the config perspective. Like if Filter is specified, we check if both PreFilter and/or Filter extensions are implemented.

Author: > One important point here is that this multipoint approach doesn't solve the problem of refactoring plugins to add new extension points since we will continue to support the fine grained configuration

Could you explain this a bit more? If a plugin is refactored to implement a new extension point, this multipoint approach will automatically register it in the new extension point as well.

Reviewer: The multipoint will handle it, but we are still supporting the original configuration where admins could explicitly list the set of extensions, and so if we refactor a plugin that initially implemented only Score to PreScore+Score, they will be surprised that their configuration which included enabling only the Score extension is no longer sufficient.

Author: we should take a position on recommending the new configuration to administrators since it addresses problems like that. If they are still using the old one while an alternative exists, then I think known issues like that can be expected.

I'm not opposed to changing some of the old config like you suggested, but I think it's not relevant to multipoint since it is solving that problem for itself

Reviewer: I don't think this proposal is totally irrelevant. The core of the MultiPoint proposal is to preserve the old configuration approach, right? If we didn't want admins to use the old approach, we wouldn't introduce MultiPoint, we would rather just force it all the time as I initially suggested in https://github.com/kubernetes/kubernetes/issues/102303. One of the motivations for https://github.com/kubernetes/kubernetes/issues/102303 was actually allowing to re-factor plugins, see quote below 

> If we decide to re-factor a plugin to add more extension points, this becomes a backward incompatible change since adding or removing an extension point may break existing configurations (see this example Scheduler(NodeVolumeLimits plugin): filter volumes of new pod in prefilter phase #102235 (comment)).

Making a recommendation to only use MultiPoint isn't sufficient to introduce changes that break the old config approach since it is still officially supported. 

In any case, I think such refactorings are usually limited to doing pre-calculations, hence the suggestion to fuse all Pre* extensions just like we did with Normalize and I feel this should be part of this proposal (simplifying plugin config), but I am not going to veto the proposal without doing this if others think it is not the time to pursue it or significantly increases the scope of this enhancement. @Huang-Wei @alculquicondor  wdyt?

Reviewer 2: There is also one semantic problem:
PreFilter can be used to mark a Pod Unschedulable at the cluster level (without the need to check any Nodes). So in a way PreFilter might exist on its own.

Maybe PreScore doesn't make sense on its own, but it's kind of awkward if PreScore is merged with Score but PreFilter is not merged with Filter.

PostFilter is supposed to be separate from Filter in any case. This is not in consideration for merging.

Reviewer: I think enlarging the scope is fine, shouldn't ideally be a blocker. I do think it is something we should seriously consider, may be in a v1beta4 :)

Reviewer: > PreFilter can be used to mark a Pod Unschedulable at the cluster level (without the need to check any Nodes). So in a way PreFilter might exist on its own.

That is fine, any plugin listed under "Filter" can implement either both or one of them.

code that utilizes the factory function stored in the `frameworkImpl` struct)

This code is similar to the existing [`updatePluginList`](https://github.com/kubernetes/kubernetes/blob/dc079acc2be603fbc2065e2748c79bb5e2c3453d/pkg/scheduler/framework/runtime/framework.go#L418)
function except for 2 notable differences:

1. It does not return an error if a plugin doesn't implement an extension point, it just continues
2. It is only looping through the MultiPoint config field for each internal extension point

With this implementation, there is no option for plugins or users to define a subset of extension 
points to enable or disable should we deprecate the existing regular extension points. This is because 
the plugins themselves do not have control over when their functions are called, and we have cast them 
into each extension point. At that point the framework, not the plugin, takes over.

Disabling of individual subsets could be preserved using modified config semantics as below:

|   |   |enabled: xyz|disabled: xyz|enabled: *|disabled: *|
|---|---|------------|-------------|----------|-----------|

> |default|regular|reorder plugins/set weight|disable plugin xyz|no-op|disable all plugins of the point|\

Reviewer 3: To make the wording consistent:

disable plugin xyz -> disable plugin xyz for this extension point

> |   |multi-point|enable plugin xyz for all extension points|disable plugin xyz for all extension points|no-op|disable all in-tree plugins for all extension points|

Reviewer 3: It seems we prefer to use "default" to represent in-tree & defaulted plugins, so let's replace all the occurrences if "in-tree"/"out-of-tree" to "default"/"non-default".

### Rollout, Upgrade and Rollback Planning

###### How can a rollout or rollback fail? Can it impact already running workloads?

Misconfigured plugins could cause the scheduler to CrashLoop as it fails to initialize the framework.

This does not have any effect on running workloads.

###### What specific metrics should inform a rollback?

Scheduler CrashLoops could indicate a failure to properly initialize the plugin registry

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

This will be manually tested

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

No

### Monitoring Requirements

###### How can an operator determine if the feature is in use by workloads?

This feature has no functional change to workloads. The scheduler's functionality can be
measured by its throughput and startup time.

###### How can someone using this feature know that it is working for their instance?

- [ ] Events
  - Event Reason: 
- [ ] API .status
  - Condition name: 
  - Other field: 
- [x] Other (treat as last resort)

>   - Details: Pods are scheduled as expected indicates functioning kube-scheduler

Reviewer 4: NodeName being assigned?
Pod not being marked as unschedulable?

Author: This is tough to explicitly define, because pods could still be marked as unschedulable (or no NodeName assigned) in a functioning scheduler. Though I will add these as indicators since a high volume of them is suspect

Reviewer 2: which means that people can also look at the `schedule_attempts_total` metric :)

Author: Thanks Aldo! I'll add that too

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
  - schedulerName: default-scheduler
    plugins:
      multiPoint:
        enabled:
        - name: "XYZ"
          weight: 3
```

#### Story 2

As a cluster administrator, I wish to enable plugin `XYZ` for only *some* of its applicable extension 
points (ie `Filter` but *not* `Score`). I still do not care about extension points such 
as `PreFilter`. I want to have control over which subsets of extension points are enabled.

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
  - schedulerName: default-scheduler
    plugins:
      multiPoint:
        enabled:
        - name: "XYZ"
      preScore:
        disabled:
        - "*"
      score:
        disabled:
        - "*"
```

#### Story 3

As a plugin developer, I want users to be able to easily enable my plugin for all applicable 
extension points (or a subset of extension points as defined either by me or my users' needs). 
I want this available with minimal code updates required from me.

### Notes/Constraints/Caveats (Optional)

> * changes should have minimal effect on existing plugins and users

Reviewer 2: I think we can remove this section.

### Risks and Mitigations

* out-of-tree plugins may be developed by anyone, so the information that is 
passed to them and received from them should be kept to a minimum. However these are 
unsupported so security is ultimately up to users to be aware of.
* UX will need to be carefully examined, as a new config field can introduce confusing 
and unexpected outcomes if different combinations of options are not well-defined.

## Design Details

> 

Reviewer 2: Also part of the Design is what happens with the defaults. Today, when users export the component config (via `write-config-to`), they get a full detail of each of the extension points. Will the defaults be migrated to the `MultiPoint` extension point?

Author: This depends on at what point in the code `write-config-to` operates. If it's before framework initialization, then users would get the `MultiPoint` extension exported. If it's after the `MultiPoint` extension is unrolled, then they will get the old config.

Reviewer 2: So you are saying that the default configuration now will use `MultiPoint` instead of individual extension points.

We should write the config after unrolling.

Author: Yes, that's the case. As long as we update the default config to use MultiPoint (I had an example of this), then that's what will get printed

Reviewer 2: Sounds good. Please add this as part of the Design Details.

    plugins:
      multiPoint:
        enabled:
        - name: "PodTopologySpread" ## PreFilter/Filter/PreScore/Score
          weight: 3
        - name: "VolumeBinding" ## PreFilter/Filter/Reserve/PreBind
        - name: "PrioritySort" ## QueueSort
```

Specific sections (eg, `Score`) could still be entirely disabled like so:

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta3
kind: KubeSchedulerConfiguration
profiles:
  - schedulerName: default-scheduler
    plugins:
      multiPoint:
        enabled:
        - name: "PodTopologySpread" ## PreFilter/Filter
          weight: 3
        - name: "VolumeBinding" ## PreFilter/Filter/Reserve/PreBind
        - name: "PrioritySort" ## QueueSort
      preScore:
        disabled:
        - "*"
      score:
        disabled:
        - "*"
```
(Note that in the above, now `PodTopologySpread` will not be registered for `Score` or `PreScore`)

To do this, we must first define a new extension point `MultiPoint` in [`Plugins`](https://github.com/kubernetes/kube-scheduler/blob/bb2d43a/config/v1beta2/types.go#L154-L159)
```go
type Plugins struct {
  MultiPoint PluginSet `json:"multiPoint,omitEmpty"`
...
}
```

> This option defines the `MultiPoint` extension as transparently applicable to all plugins, using 

Reviewer 2: ```suggestion
We define the `MultiPoint` extension as transparently applicable to all plugins, using 
```

|   |multi-point|enable plugin xyz for all extension points|disable plugin xyz for all extension points|no-op|disable all in-tree plugins for all extension points|
|non-default|regular|enable plugin xyz|disable plugin xyz for this extension point|no-op|disable all plugins of this extension point|
|   |multi-point|enable plugin xyz for all extension points|no-op|no-op|no-op|

The key changes are in `out-of-tree/regular`, where now `disabled` would take effect on 
out-of-tree plugins. Currently, out-of-tree plugins are not run unless they are `enabled`, so `disabled` 
is a no-op. But to provide subset control of multipoint plugins, we could modify this to allow disabling 
a particular extension point. This would need additional code changes to the internal processing of 
`disabled` fields, as they are currently only parsed against default plugin initialization.

### Test Plan

* Unit and integration tests to ensure multiPoint plugin configs are 
translated to their expected expanded config
* Existing scheduler e2es will ensure continued functioning of the 
scheduler framework and default plugins

### Graduation Criteria

#### Alpha

> - N/A (this feature is targeted directly at beta in the `v1beta3` scheduler API)

Reviewer 2: Add a link here to the existing KEP.

config to enable or disable plugins across multiple extension points. It will be up to
the user to know if the plugin they are using implements the new extension point, however
in-tree plugins can guarantee to implement it and the existing framework error handling
should promptly alert users for any third-party plugins which do not implement it.

### User Stories (Optional)

#### Story 1

#### Story 2

### Notes/Constraints/Caveats (Optional)

<!--
What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

An example of the new user-facing profile config would be:

```yaml

> apiVersion: kubescheduler.config.k8s.io/v1beta1

Reviewer 2: v1beta3 :)

the scheduling cycle. Some extension points, like `PreFilter` and `PreScore`, are only
used for data processing or other internal functions not directly relevant to the end user.

However these "behind-the-scenes" endpoints must still be configured in the `KubeSchedulerConfiguration`
provided by the user for every plugin that uses them. For the average user this is an unnecessary
and confusing step -- they only care that the plugin works, not how it works, and so
internal processing steps are usually irrelevant to them.

Therefore, a way for users to simply toggle all of a plugin's extension points
(or a subset of the extension points as defined necessary by the plugin developer)
would remove complexity and the potential for misconfiguration.

### Goals

* Provide an option for simplified plugin config from the end-user's perspective
* Enable this config in all in-tree default scheduler plugins

### Non-Goals

* Refactor or introduce any breaking changes to the scheduling framework

## Proposal

A new extension point, `MultiPoint` (final name TBD) will be made available to
users and developers which handles plugin registration for multiple extension
points simultaneously.

> From the developer's perspective, this extension point will be an optional interface to

Reviewer 2: Is this the option we are going with?
Even if we are, these details could go as part of the "Design Details" section.

implement in their plugin's code. This approach has several benefits:

* **Does not break existing plugins**: Developers will not need to react to this change in
order for their plugins to be configurable
* **Code organization**: A specific extension point provides an accepted pattern
and precedent to follow keeping the plugin ecosystem consistent, while also simplifying
necessary documentation and examples needed from the scheduling maintainers.
* **Consistent with other extension points**: This design can follow the established
pattern for extension points within the scheduling framework and meet all of the definitions
therein. This makes it easy to understand within the existing constructs of the framework.

From the user's perspective, this will be a new (optional) field in the scheduler's
config to enable or disable plugins across multiple extension points. It will be up to
the user to know if the plugin they are using implements the new extension point, however
in-tree plugins can guarantee to implement it and the existing framework error handling
should promptly alert users for any third-party plugins which do not implement it.

### User Stories (Optional)

#### Story 1

> As a user, I wish to enable plugin `XYZ` for all applicable extension points (ie, `Filter` 

Reviewer 2: nit: s/user/cluster administrator

and `Score`). I do not care about internal implementation or extension points such as 
`PreFilter`, nor do I want to have to update my config should plugin `XYZ` change its 
implementation within the scheduling framework to use different extension points.

#### Story 2

As a user, I wish to enable plugin `XYZ` for only *some* of its applicable extension 
points (ie `Filter` but *not* `Score`). I still do not care about extension points such 

> as `PreFilter`. I want to have control over which subsets of extension points are enabled.

Reviewer 2: Can you include the sample YAML to achieve each of these stories?

Author: For this story (subsets) the sample YAML is dependent on which implementation we choose. Option 1 defines allows subset disabling through plugin args while option 2 would use the existing regular extension points

Reviewer 2: It sounds like Option 2 would detach the user intent from what the developer provides. That is a big benefit for out-of-tree plugins. It also reduces effort for us (kube-scheduler) to maintain disablement arguments for each plugin.

```

Specific sections (eg, `Score`) could still be entirely disabled like so:

```yaml
apiVersion: kubescheduler.config.k8s.io/v1beta1
kind: KubeSchedulerConfiguration
profiles:
  - schedulerName: default-scheduler
    plugins:
      multiPoint:
        enabled:
        - name: "PodTopologySpread" ## PreFilter/Filter
          weight: 3
        - name: "VolumeBinding" ## PreFilter/Filter/Reserve/PreBind
        - name: "PrioritySort" ## QueueSort
      score:
        disabled:
        - "*"
```
(Note that in the above, now `PodTopologySpread` will not be registered for `Score` or `PreScore`)

(Also note, the semantics of doing such an approach may vary depending on the implementation design 
chosen to add multiPoint plugins)

To do this, we must first define a new extension point `MultiPoint` in [`Plugins`](https://github.com/kubernetes/kube-scheduler/blob/bb2d43a/config/v1beta2/types.go#L154-L159)
```go
type Plugins struct {
  MultiPoint PluginSet `json:"multiPoint,omitEmpty"`
...
}
```

From this point there are 2 implementation options. For both, we will need to support the 
current "regular" extension points at least until they are able to be deprecated.

> [**Option 1**](#Option-1)

Reviewer 2: Which option are you proposing we take? That one that we don't take should go in the "Alternatives" section.

Author: I personally prefer option 1, but I may be biased in that regard. I tried to present them here equally because it sounded like there was a good amount of support for option 2

Reviewer 3: Here lies a divergence of views. We can look at it in two personas: administrators / ops who configure the scheduler, and plugin developers.

For option 1, it enables the plugin developers the most flexibility to configure a plugin programmatically and dynamically - which may resolve some cases we haven't met so far. Also, it's possible to use `multiPoint` as the **only** way to compose scheduler config in the future. For admins, they get more work to do - in terms of configuration as well as understanding the plugin registration logic so as not to be surprised by something like plugin args cascading with enablement of a plugin. Also, this brings a challenge to explain everything precisely in layman's terms.

For option 2, we lost the flexibility of registrating a multiPoint plugin in runtime. So every config fields given by the admins are the final configuration. This also implies it's a must to keep multiPoint and other points in the future v1 config.

Generally, I'm inclined to option 2 as it defines a smaller but workable scope, and less code complexity and semantic ambiguity to maintain. Also, if the user wants to modify the scheduler config, modifying a static config yaml can be way easier than modifying the src code.

```go
// frameworkImpl is the component responsible for initializing and running scheduler
// plugins.
type frameworkImpl struct {
...
	pluginsMap           map[string]config.Plugin // <--new
	multiPointPlugins    []framework.MultiPointPlugin // <--new
	scorePluginWeight    map[string]int
	queueSortPlugins     []framework.QueueSortPlugin
	preFilterPlugins     []framework.PreFilterPlugin
	filterPlugins        []framework.FilterPlugin
	postFilterPlugins    []framework.PostFilterPlugin
	preScorePlugins      []framework.PreScorePlugin
	scorePlugins         []framework.ScorePlugin
	reservePlugins       []framework.ReservePlugin
	preBindPlugins       []framework.PreBindPlugin
	bindPlugins          []framework.BindPlugin
	postBindPlugins      []framework.PostBindPlugin
	permitPlugins        []framework.PermitPlugin
...
}
```

> * Modify [`getExtensionPoints()`](https://github.com/kubernetes/kubernetes/blob/dc079acc2be603fbc2065e2748c79bb5e2c3453d/pkg/scheduler/framework/runtime/framework.go#L119) to return a the `MultiPoint` extensions as well, so that the factory functions for these get registered

Reviewer 2: This KEP is going into the weeds by including too much code. It should definitely go away once the option is moved to "Alternatives".

Author: I specifically included very detailed code to show the exact steps each implementation will require in order to function as designed. I added more in the last update to answer some questions about how aspects like API defaulting will work. But I agree that once we have answered all questions and move forward with an implementation that this can definitely be cleaned up. I just didn't want to leave any "surprises" 🙂

regular extension points:

|   |   |enabled: xyz|disabled: xyz|enabled: *|disabled: *|
|---|---|------------|-------------|----------|-----------|
|in-tree|regular|reorder plugins/set weight|disable plugin xyz|no-op|disable all plugins of the point|\
|   |multi-point|enable plugin xyz for all extension points|disable plugin xyz for all extension points|no-op|disable all in-tree plugins for all extension points|
|out-of-tree|regular|enable plugin xyz|no-op|no-op|no-op|
|   |multi-point|enable plugin xyz for all extension points|no-op|no-op|no-op|

Individual plugins may be disabled using `PluginArgs` to pass a config option for which extension 
points the factory functions should register.

#### Option 2

Modify [`f.pluginsNeeded`](https://github.com/kubernetes/kubernetes/blob/dc079acc2be603fbc2065e2748c79bb5e2c3453d/pkg/scheduler/framework/runtime/framework.go#L1164) 
to additionally parse MultiPoint plugins so the function 
properly builds the map of `pluginName->New function`. Otherwise, we will not have a 
reference to any plugins that are only enabled through MultiPoint later when we 
try to unroll them to their individual extension points.

```go

> func (f *frameworkImpl) pluginsNeeded(plugins *config.Plugins) map[string]config.Plugin {

Reviewer 2: Leave code for code reviews :)

In most of these cases, it's enough to point to the functions were you need to inject new code and describe the intention of the change.

Reviewer 2: I still stand that we don't need the code in the KEP

Author: +1 removed

  // to check double registration
  set := sets.NewString()
  for _, plugin := range *e.plugins {
    set.Insert(plugin.Name)
  }
  
  for _, ep := range profile.Plugins.MultiPoint.Enabled {
    pg, ok := pluginsMap[ep.Name]
    if !reflect.TypeOf(pg).Implements(pluginType) {
      continue
    }
    
    if !ok {
      return fmt.Errorf("%s %q does not exist", pluginType.Name(), ep.Name)
    }
    
    if set.Has(ep.Name) {
      return fmt.Errorf("plugin %q already registered as %q", ep.Name, pluginType.Name())
    }
    
    set.Insert(ep.Name)
    
    newPlugins := reflect.Append(plugins, reflect.ValueOf(pg))
    plugins.Set(newPlugins)
  }
}
```

This code is similar to the existing [`updatePluginList`](https://github.com/kubernetes/kubernetes/blob/dc079acc2be603fbc2065e2748c79bb5e2c3453d/pkg/scheduler/framework/runtime/framework.go#L418)
function except for 2 notable differences:

1. It does not return an error if a plugin doesn't implement an extension point, it just continues
2. It is only looping through the MultiPoint config field for each internal extension point

With this implementation, there is no option for plugins or users to define a subset of extension 

> points to enable or disable should we deprecate the existing regular extension points. This is because 

Reviewer: The users can still select a subset of extensions to enable for a specific plugin using the old configuration semantics.

Author: `there is no option for plugins or users to define a subset of extension 
points to enable or disable should we deprecate the existing regular extension points`, this line is referring to the case where we potentially remove the old configuration semantics (which was briefly mentioned)

> the plugins themselves do not have control over when their functions are called, and we have cast them 

Reviewer: That is actually better, we want the config api to be narrow and concise.

Author: just explaining why this means users will not have subset control if the old configuration is ever removed

into each extension point. At that point the framework, not the plugin, takes over.

Disabling of individual subsets could be preserved using modified config semantics as below:

|   |   |enabled: xyz|disabled: xyz|enabled: *|disabled: *|
|---|---|------------|-------------|----------|-----------|

> |in-tree|regular|reorder plugins/set weight|disable plugin xyz|no-op|disable all plugins of the point|\

Reviewer 2: Are they all no-ops or should some of them lead to validation errors?

|   |multi-point|enable plugin xyz for all extension points|disable plugin xyz for all extension points|no-op|disable all in-tree plugins for all extension points|

> |out-of-tree|regular|enable plugin xyz|disable plugin xyz for this extension point|no-op|disable all plugins of this extension point|

Reviewer 2: I don't think the distinction is between in-tree and out-of-tree, but enabled-by-default and disabled-by-default.

(We have in-tree plugins that are disabled by default)
